### PR TITLE
Shell32 library

### PR DESCRIPTION
**Pull request by arlm**
I added KnowFolders related funcitions and constants

### CODE REVIEW

File: src/Shell32.Desktop/Shell32.Helpers.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            IntPtr pidl;
+#pragma warning disable CS0618 // Type or member is obsolete
+            if (SHGetFolderLocation(IntPtr.Zero, csidl, IntPtr.Zero, 0, out pidl) != HResult.Code.S_OK)
+#pragma warning restore CS0618 // Type or member is obsolete
+            {
+                throw new Win32Exception();
+            }
+
+            const int bufferSize = 260;
+            char* szPath = stackalloc char[bufferSize]; // max class name length
```
Comment (vbfox) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: class name ?

File: src/Shell32.Desktop/Shell32.Helpers.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        public static unsafe string GetKnownFolderID(Guid rfid)
+        {
+            IntPtr pidl;
+             if (SHGetKnownFolderIDList(rfid, 0, IntPtr.Zero, out pidl) != HResult.Code.S_OK)
+            {
+                throw new Win32Exception();
+            }
+
+            const int bufferSize = 260;
+            char* szPath = stackalloc char[bufferSize]; // max class name length
```
Comment (vbfox) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: Same :grinning:

File: src/Shell32.Desktop/Shell32.Helpers.cs (hunk 1 above)
Comment (arlm) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: Again?!

File: src/Shell32.Desktop/Shell32.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            out IntPtr pszPath);
+
+        [DllImport(nameof(Shell32), SetLastError = true)]
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for SHGetKnownFolderIDList")]
+        public static extern unsafe HResult SHGetFolderLocation(
+            IntPtr hwndOwner,
+            CSIDL nFolder,
+            IntPtr hToken,
+            int dwReserved,
+            out IntPtr pidl);
```
Comment (vbfox) [inline src/Shell32.Desktop/Shell32.cs:95]: Isn't the pidl a pointer to a known type ?

_Edit_: Seem to be a pointer to [ITEMIDLIST](https://msdn.microsoft.com/en-us/library/windows/desktop/bb773321%28v=vs.85%29.aspx) that you already add :)

File: src/Shell32.Desktop/Shell32+SHITEMID.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            /// <summary>
+            /// The size of identifier, in bytes, including cb itself.
+            /// </summary>
+            ushort cb;
+
+            /// <summary>
+            /// A variable-length item identifier.
+            /// </summary>
+            public byte[] abID;
```
Comment (vbfox) [inline src/Shell32.Desktop/Shell32+SHITEMID.cs:28]: To be able to get a pointer to the structure, it will need to be a `byte*`.

File: src/Shell32.Desktop/Shell32.Helpers.cs (hunk 2 above)
Comment (vbfox) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: Same :grinning:

**Comment by vbfox** [comment]
Except for the `IntPtr` -> pointer the rest LGTM

File: src/Shell32.Desktop/Shell32.cs (hunk 3 above)
Comment (arlm) [inline src/Shell32.Desktop/Shell32.cs:95]: Sorry, I tried hard to make it better, but couldn't. I indeed added ITEMIDLIST to that purpose but the out thingy got me nuts, I tried the FriendlyAttribute.Array without success because it complained about being a managed type pointer. I thought of Marshalling int on an Helper function, but had no success also.

Researching on that I found out that most people just use it as IntPtr and pass it along to other funcions instead of trying to Marshall it. The only reference on Marshalling I found was byte[] being "hand" readen and written with auxiliary functions and thought it was not worth publicizing as a blatant hack like this. So I just published the native functions as you see and added a Helper that calls two functions and extracts the string representation of this path instead.

Can you help me on Marshalling it?

File: src/Shell32.Desktop/Shell32+SHITEMID.cs (hunk 4 above)
Comment (arlm) [inline src/Shell32.Desktop/Shell32+SHITEMID.cs:28]: This is not a structure. abID is indeed a byte array of the identifier as it is described on [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/bb759800%28v=vs.85%29.aspx).

If you mean, to be able to make a ITEMIDLIST\* on the function I will need to make a byte\* here, ok.

I did a pointer on a struct on the User32 pull-request (MOUSEEVT) and had no success on using the IntPtr accessor. All my pull-requests are backed by in-house PInvokes that I had on the system that I am migrating to use this project. Unfortunately I could not solve the void\* on that structure for now. I really disliked this practice. Sorry.

File: src/Shell32.Desktop/Shell32+ITEMIDLIST.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        /// Contains a list of item identifiers.
+        /// </summary>
+        /// <remarks>Used by <see cref="SHGetKnownFolderIDList"/> and <see cref="SHGetFolderLocation"/></remarks>
+        [StructLayout(LayoutKind.Sequential)]
+        public struct ITEMIDLIST
+        {
+            /// <summary>
+            /// A list of item identifiers.
+            /// </summary>
+            [MarshalAs(UnmanagedType.Struct)]
```
Comment (AArnott) [inline src/Shell32.Desktop/Shell32+ITEMIDLIST.cs:23]: `MarshalAs` usually makes the containing struct no longer pinnable. Why is this necessary?

Comment (arlm) [inline src/Shell32.Desktop/Shell32+ITEMIDLIST.cs:23]: Sorry but now I have no idea of what you are talking about. Will research on pinnable structs on C# and Marshalling to understand what you are asking.

File: src/Shell32.Desktop/Shell32+KNOWNFOLDERID.cs [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+            /// <remarks>Maps to %USERPROFILE%\Videos</remarks>
+            public static readonly Guid FOLDERID_Videos = new Guid("18989B1D-99B5-455B-841C-AB7C74E4DDFC");
+
+            /// <summary>Videos</summary>
+            /// <remarks>Maps to %APPDATA%\Microsoft\Windows\Libraries\Videos.library-ms</remarks>
+            public static readonly Guid FOLDERID_VideosLibrary = new Guid("491E922F-5643-4AF4-A7EB-4E7A138D8174");
+
+            /// <summary>Windows</summary>
+            /// <remarks>Maps to %windir%</remarks>
+            public static readonly Guid FOLDERID_Windows = new Guid("F38BF404-1D43-42F2-9305-67DE0B28FC23");
```
Comment (AArnott) [inline src/Shell32.Desktop/Shell32+KNOWNFOLDERID.cs:523]: Tip: Constructing `Guid` objects based on integers rather than parsing strings [is 42X faster](https://twitter.com/aarnott/status/701653964823990273). For a large set of static fields that could help startup perf for a consuming application. 
Not necessary for this PR, but sometime I'd like to write a tool that will convert all these to their faster variety.

File: src/Shell32.Desktop/Shell32+KNOWN_FOLDER_FLAG.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        /// <summary>Specify special retrieval options for known folders.</summary>
+        /// <remarks>These values supersede CSIDL values, which have parallel meanings.</remarks>
+        [Flags]
+        public enum KNOWN_FOLDER_FLAG : uint
+        {
+            /// <summary>
+            /// Define no flags.
+            /// </summary>
+            KF_FLAG_NONE = 0x00000000,
```
Comment (AArnott) [inline src/Shell32.Desktop/Shell32+KNOWN_FOLDER_FLAG.cs]: None

File: src/Shell32.Desktop/Shell32+SHITEMID.cs (hunk 4 above)
Comment (AArnott) [inline src/Shell32.Desktop/Shell32+SHITEMID.cs:28]: +1 to @vbfox's point of it breaking pinnable structs to use `byte[]`
But perhaps more importantly, I don't think your use of `byte[]` works here. That will result in an 8-byte struct, where the last 4 bytes are a pointer to the actual array. But my reading of the MSDN doc for this structure is that the array itself is inline with the struct itself. Basically this struct defines the structure of an array that has a leading `ushort` to specify its length. We have a _few_ examples of inline arrays in structs already in this library, although I'm not sure where they are at. The fact that this one is variable length might be a new twist though. 
Is this why you added the `[MarshalAs(Struct)]` attribute on your use of this structure? Perhaps that does the magic you need, but it's unfortunate that it makes both structures un-pinnable, and requires special knowledge of this struct in every use of it.

Comment (AArnott) [inline src/Shell32.Desktop/Shell32+SHITEMID.cs:28]: > Unfortunately I could not solve the void\* on that structure for now. I really disliked this practice. Sorry.

Let's work through it. Our goals are certainly not to make the API more difficult -- just more flexible and performant. We generally have patterns we can apply to keep things simple even when pointers are present.

Comment (AArnott) [inline src/Shell32.Desktop/Shell32+SHITEMID.cs:28]: I found [the doc on inlining arrays to structs](https://msdn.microsoft.com/en-us/library/zycewsya.aspx) but it doesn't discuss dynamic sized arrays. I'm pretty sure we've faced this before... I just don't remember how we solved it.

File: src/Shell32.Desktop/Shell32.Helpers.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+            const int bufferSize = 260;
+            char* szPath = stackalloc char[bufferSize]; // max path length
+
+            if (!SHGetPathFromIDList(pidl, szPath))
+            {
+                throw new Win32Exception();
+            }
+
+            Marshal.FreeCoTaskMem(pidl);
```
Comment (AArnott) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: It looks like you'll leak `pidl` unless you add try/finally around your allocation and freeing. If you throw `Win32Exception`, you'll leak it.

File: src/Shell32.Desktop/Shell32.Helpers.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            return new string(szPath, 0, bufferSize);
+        }
+
+        public static unsafe string GetKnownFolderID(Guid rfid)
+        {
+            IntPtr pidl;
+             if (SHGetKnownFolderIDList(rfid, 0, IntPtr.Zero, out pidl) != HResult.Code.S_OK)
```
Comment (AArnott) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: Format Document to fix whitespace/indentation issues.

File: src/Shell32.Desktop/Shell32.Helpers.cs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            Marshal.FreeCoTaskMem(pszPath);
+            return path;
+        }
+
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for GetKnownFolderID")]
+        public static unsafe string GetFolderLocation(CSIDL csidl)
+        {
+            IntPtr pidl;
+#pragma warning disable CS0618 // Type or member is obsolete
+            if (SHGetFolderLocation(IntPtr.Zero, csidl, IntPtr.Zero, 0, out pidl) != HResult.Code.S_OK)
```
Comment (AArnott) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: I tend to prefer accepting all success codes rather than only `S_OK`. It looks like in at least some cases the MSDN docs state S_OK is the only success code returned, but unless you've verified that for every function, can you use this syntax instead?

``` csharp
if (SHGetFolderLocation(IntPtr.Zero, csidl, IntPtr.Zero, 0, out pidl).Failed)
```

IMO it's easier to read anyway.

File: src/Shell32.Desktop/Shell32.cs (hunk 3 above)
Comment (AArnott) [inline src/Shell32.Desktop/Shell32.cs:95]: I'm interested in trying my hand at solving this. @vbfox is of course welcome to beat me to it. I won't get to it till tonight or tomorrow.

**Comment by AArnott** [comment]
Why is your `src/Shell32.Desktop/Shell32.exports.txt` file empty? It should have a bunch of function names in it. The `AddNewLibrary.ps1` script generally takes care of this for you. Did it malfunction? 
Could you perhaps try directly running `CreateExportsTxtFile.ps1` to see if it can create it properly?

File: src/Shell32.Desktop/Shell32+ITEMIDLIST.cs (hunk 5 above)
Comment (AArnott) [inline src/Shell32.Desktop/Shell32+ITEMIDLIST.cs:23]: You mentioned you couldn't use pointers in your P/Invoke method signatures for this struct. This is why. You can't use pointers to structs that have `MarshalAs` attributes in them. But I can help investigate the overall solution as discussed below.

File: src/Shell32.Desktop/Shell32+KNOWNFOLDERID.cs (hunk 6 above)
Comment (arlm) [inline src/Shell32.Desktop/Shell32+KNOWNFOLDERID.cs:523]: A tool would be great. It could be integrated on the build on the CodeGen phase, just changing the code instead of generating. hehehe.

File: src/Shell32.Desktop/Shell32.Helpers.cs (hunk 8 above)
Comment (arlm) [inline src/Shell32.Desktop/Shell32.Helpers.cs]: Good point.

**Comment by arlm** [comment]
Strange. It seems that it failed. But the AddNewLibrary run.

File C:...\pinvoke\templates\CreateExportsTxtFile.ps1 cannot be loaded because running scripts is disabled on this system. For more information, see 
about_Execution_Policies at http://go.microsoft.com/fwlink/?LinkID=135170.
    + CategoryInfo          : SecurityError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : UnauthorizedAccess

**Comment by arlm** [comment]
Is there any other way to populate it?

File: src/Shell32.Desktop/Shell32+KNOWNFOLDERID.cs (hunk 6 above)
Comment (AArnott) [inline src/Shell32.Desktop/Shell32+KNOWNFOLDERID.cs:523]: That's not too far from what I was thinking. A Roslyn analyzer could find these and offer an automatic code fixer for it.

**Comment by AArnott** [comment]
You could run this powershell command to unblock it, I think:
`set-executionpolicy remotesigned`

File: src/Shell32.Desktop/Shell32+SHITEMID.cs (hunk 4 above)
Comment (vbfox) [inline src/Shell32.Desktop/Shell32+SHITEMID.cs:28]: Ah yes this isn't doable with the marshaler anyway. it should be done by declaring a fixed 1 element array and then allocating the structure by hand if we need to allocate it (Exactly as the C++ code does)

I don't know if we need to get values in and out of this structure (Maybe it's declared but enough API are available that it's never needed to access it). If it is necessary it would look like that :

``` csharp
public unsafe partial class Shell32
{
    /// <summary>
    /// Defines an item identifier.
    /// </summary>
    /// <remarks>Used by <see cref="ITEMIDLIST"/></remarks>
    [StructLayout(LayoutKind.Sequential)]
    public struct SHITEMID
    {
        public const short FixedPartSize = 2; // sizeof(ushort)

        /// <summary>
        /// The size of identifier, in bytes, including cb itself.
        /// </summary>
        public short cb;

        /// <summary>
        /// A variable-length item identifier.
        /// </summary>
        public fixed byte abID[1];

        public byte[] GetValue()
        {
            var length = cb - FixedPartSize;
            var result = new byte[length];
            fixed (byte* valuePtr = abID)
            {
                Marshal.Copy((IntPtr)valuePtr, result, 0, length);
            }
            return result;
        }

        public static SHITEMID* Create(short size)
        {
            var totalSize = FixedPartSize + size;
            var result = (SHITEMID*)Marshal.AllocHGlobal(totalSize);
            result->cb = (short)totalSize;
            return result;
        }
    }
}
```

File: src/Shell32.Desktop/Shell32.cs (hunk 3 above)
Comment (vbfox) [inline src/Shell32.Desktop/Shell32.cs:95]: I don't think i'll have time to do it before this weekend but with something approaching what I posted in my comment to `SHITEMID`, `ITEMIDLIST` itself shouldn't be problematic.

Regarding pointers, I don't know how @AArnott see it but I see this library as an attempt to get the correct types as much as we can, `SafeHandle` for opaque types having a custom destruction function, pointers instead of `IntPtr`... 
That's more complex than what is present on pinvoke.net, and there is often a big step of refactorings when importing pinvokes from an existing code base. The target isn't to have a NuGet version of pinvoke.net (I still wonder why they never did it, it would be relatively simple to create) but to work on getting a clean API surface that resemble as much as possible the C version, adding helpers to ease C# consumption.

**Comment by AArnott** [comment]
I'm working on this PR now to resolve some issues we've talked about.
Also, the `ILFree` method is missing, which evidently is a requirement when anyone calls `SHGetKnownFolderIDList`

**Comment by AArnott** [comment]
Well, I've come away convinced that pointers were definitely important to retain here. The rules for these functions and how to free the memory they allocate are diverse, and there's no way the .NET Marshaler could handle it all automatically. It must be done with native pointers (or IntPtr) rather than marshaled so that the original pointer is available to free later. Helper methods can help keep things sane for the caller though.
I'm pretty sure there were memory leaks in the original PR.

**Comment by AArnott** [comment]
`SetLastError=true` is inappropriate, because these functions don't document to use it.
Also you seem to be treating error codes as if they were Win32 errors by throwing `Win32Exception`, which is inappropriate because these functions return HRESULTs.

**Comment by AArnott** [comment]
@vbfox please consider looking at arlm/pinvoke#2 to review the changes I'm proposing to this PR.

**Comment by AArnott** [comment]
But before we can merge this PR, #248 needs to be fixed to avoid introducing a bunch of build warnings.

**Comment by vbfox** [comment]
@AArnott  arlm/pinvoke#2 LGTM, nice that we can have tests for that. I don't know if the manual methods to change / get the value under an `SHITEMID` will be useful to anyone but the code LGTM.

(Make me think that I could create a NuGet version of my answer on how to open explorer with files selected http://stackoverflow.com/a/3578581/46594 now that we are adding shell functions)

**Comment by AArnott** [comment]
Thanks.
Yes, I meant to add tests for my `SHITEMID` accessor methods. But I also found in debugging that the buffer appears to not be intended for direct consumption, so I can probably just remove them all.

**Comment by AArnott** [comment]
I've removed the accessor methods and added doc comments warnings folks to always use pointers (or IntPtr) when handling these structs.

**Comment by AArnott** [comment]
Thanks for your contribution, @arlm!
